<?xml version="1.0"?>
<robot name="prime" xmlns:xacro="http://www.ros.org/wiki/xacro">
<xacro:property name="wheel_radius" value="0.1"/>
<xacro:property name="wheel_thickness" value="0.1"/>
<xacro:property name="chassis_length" value="0.6"/>
<xacro:property name="chassis_width" value="0.3"/>
<xacro:property name="wheel_offset_y" value="${(chassis_width/2)+(wheel_thickness/2)}"/>
<xacro:property name="chassis_mass" value="0.2"/>
<xacro:property name="wheel_mass" value="0.02"/>

<material name="orange">
<color rgba="1 0.3 0.1 1"/>
</material>

<link name="chassis">
<visual>
<origin xyz="0 0 0" rpy="0 0 0"/>
<geometry>
<box size="${chassis_length} ${chassis_width} 0.1"/>
</geometry>
<material name="orange"/>
</visual>
<collision>
<geometry>
<box size="${chassis_length} ${chassis_width} 0.1"/>
</geometry>
<gazebo>
<mu1>0.8</mu1>
<mu2>0.2</mu2>
</gazebo>
</collision>
<inertial>
<mass value="${chassis_mass}"/>
<inertia ixx="1.0" ixy="0.0" iyy="1.0" izz="1.0" iyz="0.0" ixz="0.0"/>
</inertial>
</link>

<link name="wheel_1">
<visual>
<origin xyz="0 0 0" rpy="0 0 0"/>
<geometry>
<cylinder radius="${wheel_radius}" length="${wheel_thickness}"/>
</geometry>
<material name="orange"/>
</visual>
<collision>
<geometry>
<cylinder radius="${wheel_radius}" length="${wheel_thickness}"/>
</geometry>
<gazebo>
<mu1>0.8</mu1>
<mu2>0.2</mu2>
</gazebo>
</collision>
<inertial>
<mass value="${wheel_mass}"/>
<inertia ixx="0.1" ixy="0.0" iyy="0.1" iyz="0.0" izz="0.1" ixz="0.0"/>
</inertial>
</link>

<link name="wheel_2">
<visual>
<origin xyz="0 0 0" rpy="0 0 0"/>
<geometry>
<cylinder radius="${wheel_radius}" length="${wheel_thickness}"/>
</geometry>
</visual>
<collision>
<geometry>
<cylinder radius="${wheel_radius}" length="${wheel_thickness}"/>
</geometry>
<gazebo>
<mu1>0.8</mu1>
<mu2>0.2</mu2>
</gazebo>
</collision>
<inertial>
<mass value="${wheel_mass}"/>
<inertia ixx="0.1" ixy="0.0" iyy="0.1" iyz="0.0" izz="0.1" ixz="0.0"/>
</inertial>
</link>

<link name="wheel_3">
<visual>
<origin xyz="0 0 0" rpy="0 0 0"/>
<geometry>
<cylinder radius="${wheel_radius}" length="${wheel_thickness}"/>
</geometry>
</visual>
<collision>
<geometry>
<cylinder radius="${wheel_radius}" length="${wheel_thickness}"/>
</geometry>
<gazebo>
<mu1>0.8</mu1>
<mu2>0.2</mu2>
</gazebo>
</collision>
<inertial>
<mass value="${wheel_mass}"/>
<inertia ixx="0.1" ixy="0.0" iyy="0.1" iyz="0.0" izz="0.1" ixz="0.0"/>
</inertial>
</link>

<link name="wheel_4">
<visual>
<origin xyz="0 0 0" rpy="0 0 0"/>
<geometry>
<cylinder radius="${wheel_radius}" length="${wheel_thickness}"/>
</geometry>
</visual>
<collision>
<geometry>
<cylinder radius="${wheel_radius}" length="${wheel_thickness}"/>
</geometry>
</collision>
<inertial>
<mass value="${wheel_mass}"/>
<inertia ixx="0.1" ixy="0.0" iyy="0.1" iyz="0.0" izz="0.1" ixz="0.0"/>
</inertial>
</link>


<joint name="first_wheel_joint" type="revolute">
<origin xyz="${chassis_length/2} ${-wheel_offset_y} 0" rpy="1.57 0 0"/>
<parent link="chassis"/>
<child link="wheel_1"/>
<limit lower="-3.14" upper="3.14" effort="10" velocity="10"/>
<axis xyz="0 0 1"/>
</joint>

<joint name="second_wheel_joint" type="revolute">
<origin xyz="${chassis_length/2} ${wheel_offset_y} 0" rpy="1.57 0 0"/>
<parent link="chassis"/>
<child link="wheel_2"/>
<limit lower="-3.14" upper="3.14" effort="10" velocity="10"/>
<axis xyz="0 0 1"/>
</joint>

<joint name="third_wheel_joint" type="revolute">
<origin xyz="${-chassis_length/2} ${wheel_offset_y} 0" rpy="1.57 0 0"/>
<parent link="chassis"/>
<child link="wheel_3"/>
<limit lower="-3.14" upper="3.14" effort="10" velocity="10"/>
<axis xyz="0 0 1"/>
</joint>

<joint name="fourth_wheel_joint" type="revolute">
<origin xyz="${-chassis_length/2} ${-wheel_offset_y} 0" rpy="1.57 0 0"/>
<parent link="chassis"/>
<child link="wheel_4"/>
<limit lower="-3.14" upper="3.14" effort="10" velocity="10"/>
<axis xyz="0 0 1"/>
</joint>

<ros2_control name="Gazebosim" type="system">
<hardware>
<plugin>gazebo_ros2_control/GazeboSystem</plugin>
</hardware>

<joint name="first_wheel_joint">
<command_interface name="position"/>
<state_interface name="position"/>
<state_interface name="velocity"/>
</joint>

<joint name="second_wheel_joint">
<command_interface name="position"/>
<state_interface name="position"/>
<state_interface name="velocity"/>
</joint>
<joint name="third_wheel_joint">
<command_interface name="position"/>
<state_interface name="position"/>
<state_interface name="velocity"/>
</joint>
<joint name="fourth_wheel_joint">
<command_interface name="position"/>
<state_interface name="position"/>
<state_interface name="velocity"/>
</joint>

</ros2_control>
<gazebo>
<plugin name="gazebo_ros2_control" filename="libgazebo_ros2_control.so">
  <robotNamespace>/</robotNamespace>
  <parameters>$(find four_wheel_mobile_robot_navigation)/config/my_controllers.yaml</parameters>
</plugin>
</gazebo>

</robot>